src/main/java/org/springframework/data/redis/connection/lettuce/LettuceReactivePubSubCommands.java

```diff
@@ -65,20 +83,157 @@ public Mono<Void> subscribe(ByteBuffer... channels) {
+			return Mono.empty();
+		}
+
+		return doWithPubSub(commands -> commands.unsubscribe(actualUnsubscribe));
+	}
+
 	@Override
 	public Mono<Void> pSubscribe(ByteBuffer... patterns) {
 
 		Assert.notNull(patterns, "Patterns must not be null");
 
+		Target.trackSubscriptions(patterns, this.patterns); // track usage but do not limit what to subscribe to
+
 		return doWithPubSub(commands -> commands.psubscribe(patterns));
 	}
 
+	public Mono<Void> pUnsubscribe(ByteBuffer... patterns) {
+
+		Assert.notNull(patterns, "Patterns must not be null");
+
+		ByteBuffer[] actualUnsubscribe = Target.trackUnsubscriptions(patterns, this.patterns);
+
+		if (actualUnsubscribe.length == 0 && patterns.length != 0) {
+			return Mono.empty();
+		}
+
+		return doWithPubSub(commands -> commands.punsubscribe(actualUnsubscribe));
+	}
+
 	private <T> Mono<T> doWithPubSub(Function<RedisPubSubReactiveCommands<ByteBuffer, ByteBuffer>, Mono<T>> function) {
 
 		return connection.getPubSubConnection().flatMap(pubSubConnection -> function.apply(pubSubConnection.reactive()))
 				.onErrorMap(connection.translateException());
 	}
+
+	static class Target {
+
+		private static final AtomicLongFieldUpdater<Target> SUBSCRIBERS = AtomicLongFieldUpdater.newUpdater(Target.class,
+				"subscribers");
+
+		private final byte[] raw;
+
+		private volatile long subscribers;
+
+		Target(byte[] raw) {
+			this.raw = raw;
+		}
+
+		/**
+		 * Record the subscriptions to {@code targets} and store these in {@code targetMap}.
+		 *
+		 * @param targets
+		 * @param targetMap
+		 */
+		public static void trackSubscriptions(ByteBuffer[] targets, Map<ByteArrayWrapper, Target> targetMap) {
+			doWithTargets(targets, targetMap, Target::allocate);
+		}
+
+		/**
+		 * Record the un-subscriptions to {@code targets} and store these in {@code targetMap}. Returns the targets to
+		 * actually unsubscribe from if there are no subscribers to a particular target.
+		 *
+		 * @param targets
+		 * @param targetMap
+		 */
+		public static ByteBuffer[] trackUnsubscriptions(ByteBuffer[] targets, Map<ByteArrayWrapper, Target> targetMap) {
+			return doWithTargets(targets, targetMap, Target::deallocate);
+		}
+
+		static ByteBuffer[] doWithTargets(ByteBuffer[] targets, Map<ByteArrayWrapper, Target> targetMap,
+				BiFunction<ByteBuffer, Map<ByteArrayWrapper, Target>, Boolean> f) {
+
+			List<ByteBuffer> toSubscribe = new ArrayList<>(targets.length);
+
+			synchronized (targetMap) {
+				for (ByteBuffer target : targets) {
+					if (f.apply(target, targetMap)) {
+						toSubscribe.add(target);
+					}
+				}
+			}
+
+			return toSubscribe.toArray(new ByteBuffer[0]);
+		}
+
+		boolean increment() {
+			return SUBSCRIBERS.incrementAndGet(this) == 1;
+		}
+
+		boolean decrement() {
+
+			long l = SUBSCRIBERS.get(this);
+
+			if (l > 0) {
+				if (SUBSCRIBERS.compareAndSet(this, l, l - 1)) {
+					return l == 1; // return true if this was the last subscriber
+				}
+			}
+
+			return false;
```

Comment:
maybe just `return SUBSCRIBERS.getAndUpdate(this, cur -> cur > 0 ? cur - 1 : cur) == 1;`